### DISCUSSION TITLE
Mesh splitting

### DISCUSSION THREAD

**Discussion by KhaledNabilSharafeldin** [idaholab/moose#18785] [Q&A Meshing] [answered]
Hello,
I would like to ask about distributed mesh a couple of questions,
First the original problem: Using larger mesh (more elements) with 24 independent variables with 3 subapps uses more than 100gb of memory.
What I am trying to do: use mesh splitting so that it doesnt copy the whole mesh for each processor.

```
mpiexec -np 24 app-opt -i inputFile.i --split-mesh 24,48,144 --split-file foo.cpr
mpiexec -np 24 app-opt -i inputFile.i --use-split --split-file foo.cpr

```
Problem:
Error:

```
*** ERROR ***
The following error occurred in the object "PostProcessorValue", of type "NodalVariableValue".

NodalVariableValue should only be used when node renumbering is disabled.
```

This solution did nothing:
```
[Mesh]
  [./msh]
    type = FileMeshGenerator
    file = ../../test.exo
    dim = 3
    **allow_renumbering = false**
  []
  [./sidesets]
    type = SideSetsFromNormalsGenerator
    input = msh
    normals = '0 0 1
    0 0 -1
    0 1 0
    0 -1 0
    1 0 0
    -1 0 0'
    fixed_normal = true
    new_boundary = 'top1 bottom1 front1 back1 right1 left1'
  []
  [./ns1]
    type = ExtraNodesetGenerator
    input = sidesets
    new_boundary = 'nodeset1'
    nodes = '1'
  [../]
[]
```
I these variables were not essential in calculations. So I commented them out for now... but now I get this problem:

```
4[22] ./include/libmesh/remote_elem.h, line 
[120, compiled Aug  2 2021 at 10:45:35


*** ERROR ***
Error: not implemented!
```
 This happens in the subapp. 

No idea where to go from here unfortunately, would appreciate any suggestions.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Your issue is that the Mesh generators are basically ignored when a split mesh is loaded this way (using the --use-split). 
I will rework the documentation for that soon. In the mean time you could use `--distributed-mesh` and no split. Splits are typically for very very large runs where IO matters a lot. For 24 processes, you are just trying to distribute the mesh not reduce file system IO

You can still use split with a FileMeshGenerator (the documentation I ll be writing) but for the `file` argument you need to specify the 'foo.cpr' folder.

Hope this helps

Guillaume

**Reply by GiudGiud** [reply]
I re-read your post. You definitely want to be using `--distributed-mesh` instead of splitting meshes.

**Reply by KhaledNabilSharafeldin** [reply]
I'll give that a shot, but any idea on how to solve the last problem? it started showing up after testing with mesh splitting...
```
4[22] ./include/libmesh/remote_elem.h, line 
[120, compiled Aug  2 2021 at 10:45:35


*** ERROR ***
Error: not implemented!
```
I tried recompiling all 

```
make clobberall
make -j8
```
doesnt seem to work.

**Reply by GiudGiud** [reply]
We'd need a backtrace to see which function is not implemented. Libmesh is supposed to provide one, but maybe not on asserts. 
If you could run the code in a debugger and place a breakpoint on MPI_Abort that should work to get a backtrace

**Reply by KhaledNabilSharafeldin** [reply]
Thank you! I managed to get it to work, it was unrelated to the original problem. As for `--distributed-mesh` it worked like a charm. I guess splitting method would work just as well.
I have a couple of follow up questions,
1) would using more subapps require more memory?
2) when do you consider the problem large enough to use splitting, since i'm only using 24 cores for debugging/testing/setting up the simulation but have access to more computational resources. when does splitting is more advantageous than distributed mesh.
Not sure exactly the technical differences between distributed mesh and splitting it, so they might be the same thing, so I'd just like to clarify the difference. from what I understood, the mesh can get very large that it wont fit in the RAM, that's when you use pre-splitting. which might be handy to set up for the problem .

Again, thanks for the very helpful support.

**Reply by KhaledNabilSharafeldin** [reply]
Would like to add to this is that I had to add to the mesh block `allow_renumbering = false` to calculate the Nodal post-processors, and for my subapp I used `clone_master_mesh = true` in the subapp block 

Mesh:
```
[Mesh]
  [./msh]
    type = FileMeshGenerator
    file = ../../test.exo
    dim = 3
    allow_renumbering = false
  []
  [./sidesets]
    type = SideSetsFromNormalsGenerator
    input = msh
    normals = '0 0 1
    0 0 -1
    0 1 0
    0 -1 0
    1 0 0
    -1 0 0'
    fixed_normal = true
    new_boundary = 'top1 bottom1 front1 back1 right1 left1'
  []
  [./ns1]
    type = ExtraNodesetGenerator
    input = sidesets
    new_boundary = 'nodeset1'
    nodes = '1'
  [../]
  allow_renumbering = false # This is the one that worked
[]
```

Multiapp:
```
[MultiApps]
  [./cstests_sub]
    type = TransientMultiApp
    positions = '0 0 0'
    input_files = csTests_sub.i
    execute_on = 'initial timestep_end'
    clone_master_mesh = true
  [../]
[] 
```

Correct me if this is not the case for using `clone_master_mesh` or if better solution exists.

**Reply by GiudGiud** [reply]
If the subapps are using the same mesh as the main app, clone_master_mesh is great for that. 

1) Yes. More subapps will use more memory usually. In your case, if all subapps are defining variables on the same mesh as the main app, then that takes up memory. 
2) See this page for more info on splitting.  https://mooseframework.inl.gov/syntax/Mesh/splitting.html
Expensive mesh work and lots of processors (like >100) on a slow file system make splitting beneficial
